File: IntegerToEnglishWords.java
class Solution {
    String digitWordArray[] = {"", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine"};
    String tensWordArray[] = {"", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"};
    String tenSpecialArray[] = {"Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"};

    // Class defined in the test to convert an integer number into the word equivalent.
    // parameter: num is the integer number to be converted.
    // return: a string representation of the number in words.
    public String numberToWords(int num) {
        String numberWord = "";
        int[] digitArray = new int[10]; 
        int dividen = 0;
        int remainder = 0;
        int power = 0;
        
        if (num > 0) {
            // The digits are parsed from left to right to simplify the calculation of the 
            // power variable.
            for (int exp = 9; exp > -1; exp--) {
                power = (int)Math.pow(10, exp);
                dividen = num / power;
                num = num % power;
                
                if (dividen > 0) {
                    digitArray[exp] = dividen;
                }
            }
            
            // Add Billions
            if (digitArray[9] != 0) {
                numberWord = digitWordArray[digitArray[9]] + " Billion ";
            }
            
            // Add Millions
            if (digitArray[8] != 0 || digitArray[7] != 0 || digitArray[6] != 0) {
                numberWord += hundredsWords(digitArray, 8) + " Million ";
            }
            
            // Add Thousands
            if (digitArray[5] != 0 || digitArray[4] != 0 || digitArray[3] != 0) {
                numberWord += hundredsWords(digitArray, 5) + " Thousand ";
            }
            
            // Add Hundreds
            if (digitArray[2] != 0 || digitArray[1] != 0 || digitArray[0] != 0) {
                numberWord += hundredsWords(digitArray, 2);
            }
        }
        else {
            numberWord = "Zero";
        }

        return numberWord.trim();
    }  // numberToWords
 
    // This method breaks down the three digits that make up a hundreds group.
    // It is used multiple time in the millions, thousands and hundreds area.
    // parameter: digitArray is the whole array of digits parsed by the 
    //            numberToWords method.
    // parameter: index is the starting index in the digitArray to parse for
    //            the hundreds words.
    // return: The string wording of the hundreds digits.
    String hundredsWords(int[] digitArray, int index) {
        String word = "";
        
        if (digitArray[index] > 0) {
            word = digitWordArray[digitArray[index]] + " Hundred ";
        }
        
        if (digitArray[index - 1] > 1) {
            word += tensWordArray[digitArray[index - 1]] + " ";
        }
        else {  // special case
            if (digitArray[index - 1] == 1) {
                
                word += tenSpecialArray[digitArray[index - 2]] + " ";
            }
        }  
                        
        if (digitArray[index - 1] != 1) {
            word += digitWordArray[digitArray[index - 2]];
        }
        
        return word.trim();
    }  // hundredsWords
}  // end class Solution

File: IntegerToEnglishWords.cs
    public class Solution
    {
        string[] digitArray = { "Zero", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine" };  // Empty string because of zero based array
        string[] tensAmoutArray = { "", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety" };  // Empty string because of zero based array and ten amount is a special case
        string[] tensArray = { "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen" };

        static int Thousand = 1000;
        static int Million = 1000000;
        static int Billion = 1000000000;

        public string NumberToWords(int num)
        {
            string numberString = "";

            if ((num / Billion) > 0)
            {
                numberString = CreateHundredsStr(num / Billion);
                numberString += " Billion";
            }

            num %= Billion;

            if ((num / Million) > 0)
            {
                if (numberString != "")
                {
                    numberString += " ";
                }

                numberString += CreateHundredsStr(num / Million);
                numberString += " Million";
            }

            num %= Million;

            if ((num / Thousand) > 0)
            {
                if (numberString != "")
                {
                    numberString += " ";
                }

                numberString += CreateHundredsStr(num / Thousand);
                numberString += " Thousand";
            }

            num %= Thousand;

            if (num > 0)
            {
                if (numberString != "")
                {
                    numberString += " ";
                }

                numberString += CreateHundredsStr(num);
            }
            else
            {
                if (numberString == "")
                {
                    numberString = "Zero";
                }
            }

            return numberString;
        }  // NumberToWords

        private string CreateHundredsStr(int number)
        {
            string result = "";
            int digit = number / 100;

            if (digit > 0)
            {
                result = digitArray[digit] + " Hundred";

                digit = (number % 100);
            }

            digit = number % 100;

            if (digit > 9)
            {
                if (result != "")
                {
                    result += " ";
                }

                result += NumberToStr(digit);
            }
            else if (digit > 0 || (digit == 0 && result == ""))
            {
                if (result != "")
                {
                    result += " ";
                }

                result += digitArray[digit];
            }

            return result;
        }  // CreateHundredsStr

        private string NumberToStr(int digit)
        {
            string result = "";

            if (digit > 19)
            {
                result = tensAmoutArray[(digit / 10)];

                if ((digit % 10) > 0)
                {
                    result = result + " " + digitArray[(digit % 10)];
                }
            }
            else
            {  // Ten's digit is special case between 10 and 19
                if (result != "")
                {
                    result += " ";
                }

                result += tensArray[(digit % 10)];
            }

            return result;
        }  //  NumberToStr

    }  // end class Solution
